### PR TITLE
xds: update (local) configurations atomically for each LDS/RDS resource update

### LINKED ISSUE #8009: XdsNameResolver does not handle resource revoked and then recovered correctly
If LDS resource is revoked, the xds lbs will be shutdown and switch to empty service config and pick_first, but when exactly the same resource is recovered after revocation, [`shouldUpdateResult`](https://github.com/grpc/grpc-java/blob/v1.36.0/xds/src/main/java/io/grpc/xds/XdsNameResolver.java#L632-L658) is always false, and therefore `updateResolutionResult()` is never called and new service config is never generated.

### PR DESCRIPTION
**Pull request by voidzcy**
XdsNameResolver is stateful in terms of received xDS configurations (e.g., routes, httpMaxStreamDuration, clusters being exposed in service config). The full suite of configurations should always be updated whenever receiving new resource updates, including updates that revoke currently in-use resources. Reference counts for currently in-use clusters should also be cleaned up properly. Otherwise, re-receiving (after being revoked) the same resource can be treated as if the configuration never changed.

In short, all of XdsNameResolver's states should always be consistent with the most recent xDS update, including the updates for revoking resources.

-------------
Fixes #8009

### CODE REVIEW

**Review by dapengzhang0 — COMMENTED** [review]
The state of XdsNameResolve is getting more and more complex and delicate and it's easy to introduce bug :(

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java [hunk 1]
```diff
@@ -791,26 +789,56 @@ private void updateRoutes(List<VirtualHost> virtualHosts) {
...
+        xdsClient.cancelRdsResourceWatch(rdsName, routeConfigState);
+        routeConfigState = null;
+      }
+    }
+
+    /**
+     * Discovery state for RouteConfiguration resource. One instance for each Listener resource
+     * update.
+     */
+    private class RouteConfigState implements RdsResourceWatcher {
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsNameResolver.java]: Done.

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java [hunk 2]
```diff
@@ -754,13 +729,35 @@ private void updateRoutes(List<VirtualHost> virtualHosts) {
...
+          }
+        }
+        if (!hasRouter) {
+          // Fail all RPCs if a router filter is not present. Reference counts for all currently
+          // selectable clusters should be reclaimed.
+          filterChain.add(LAME_FILTER);
+          routes = Collections.emptyList();
+          clusters = Collections.emptySet();
+        }
+      }
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsNameResolver.java]: Done.

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java [hunk 3]
```diff
@@ -806,25 +806,25 @@ private void cleanUpRoutes() {
     }
 
     private void cleanUpRouteConfigState() {
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsNameResolver.java:809]: nit: method can be further renamed.

**Review by dapengzhang0 — APPROVED** [review]

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java (hunk 3 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsNameResolver.java]: Oops, missed this one. Should be fixed now.
